File: apps/backend/src/workflow-transitions.js
/**
 * Чистая бизнес-логика переходов состояний workflow.
 * Не зависит от HTTP, Supabase или Fastify — легко тестируется в изоляции.
 */

export const INTEGRATION_START_IDX = 13;

export const LAST_STEP_INDEX_BY_STAGE = Object.freeze({
  1: 6,
  2: 9,
  3: 12,
  4: 14,
});

export const TOTAL_STEPS = 15;

export function getStageStepRange(stage) {
  const normalizedStage = Number(stage || 1);
  const rangeEnd = LAST_STEP_INDEX_BY_STAGE[normalizedStage];
  if (!Number.isInteger(rangeEnd) || rangeEnd < 0) return null;

  const prevStage = normalizedStage - 1;
  const prevEnd = prevStage >= 1 ? LAST_STEP_INDEX_BY_STAGE[prevStage] : -1;
  const rangeStart = Number.isInteger(prevEnd) ? prevEnd + 1 : 0;

  return { start: rangeStart, end: rangeEnd };
}

export function buildCompletionTransition(current) {
  const currentStep = Number(current.current_step || 0);
  const currentStage = Number(current.current_stage || 1);
  const nextStepIndex = currentStep + 1;
  const stageBoundary = LAST_STEP_INDEX_BY_STAGE[currentStage] === currentStep;
  const isLastStepGlobal = nextStepIndex >= TOTAL_STEPS;

  let nextStatus = current.status || 'IN_PROGRESS';
  let nextSubstatus = current.workflow_substatus || 'DRAFT';
  let nextStage = currentStage;

  if (isLastStepGlobal) {
    nextStatus = 'COMPLETED';
    nextSubstatus = 'DONE';
  } else if (stageBoundary) {
    nextStatus = 'IN_PROGRESS';
    nextSubstatus = 'REVIEW';
    nextStage = currentStage + 1;
  } else if (nextStepIndex === INTEGRATION_START_IDX) {
    nextStatus = 'IN_PROGRESS';
    nextSubstatus = 'INTEGRATION';
  } else {
    nextStatus = 'IN_PROGRESS';
    if (nextSubstatus !== 'INTEGRATION') nextSubstatus = 'DRAFT';
  }

  return { nextStepIndex, nextStatus, nextSubstatus, nextStage };
}

export function buildRollbackTransition(current) {
  const currentStep = Number(current.current_step || 0);
  const prevIndex = Math.max(0, currentStep - 1);
  const currentSubstatus = current.workflow_substatus || 'DRAFT';

  let nextSubstatus = currentSubstatus;
  if (currentSubstatus === 'REVIEW' || currentSubstatus === 'DONE') {
    nextSubstatus = 'DRAFT';
  }

  return {
    nextStepIndex: prevIndex,
    nextStage: Number(current.current_stage || 1),
    nextStatus: 'IN_PROGRESS',
    nextSubstatus,
  };
}

export function buildReviewTransition(current, action) {
  const isApprove = action === 'APPROVE';
  const currentStage = Number(current.current_stage || 1);
  let nextStatus = current.status || 'IN_PROGRESS';
  let nextSubstatus = current.workflow_substatus || 'DRAFT';
  let nextStepIndex = Number(current.current_step || 0);
  let nextStage = currentStage;

  if (isApprove) {
    nextSubstatus = 'DRAFT';
    if (nextStepIndex === INTEGRATION_START_IDX) nextSubstatus = 'INTEGRATION';
    nextStatus = 'IN_PROGRESS';
  } else {
    nextStage = Math.max(1, currentStage - 1);
    nextStepIndex = LAST_STEP_INDEX_BY_STAGE[nextStage] ?? 0;
    nextSubstatus = 'REVISION';
    nextStatus = 'IN_PROGRESS';
  }

  return { isApprove, nextStatus, nextSubstatus, nextStepIndex, nextStage };
}

File: apps/backend-java/src/main/java/uz/reestrmkd/backend/application/WorkflowTransitions.java
package uz.reestrmkd.backend.application;

import java.util.Map;

public final class WorkflowTransitions {
    private WorkflowTransitions() {}

    public static final int INTEGRATION_START_IDX = 13;
    public static final Map<Integer, Integer> LAST_STEP_INDEX_BY_STAGE = Map.of(1, 6, 2, 9, 3, 12, 4, 14);
    public static final int TOTAL_STEPS = 15;

    public record Transition(int nextStepIndex, String nextStatus, String nextSubstatus, int nextStage) {}
    public record StageRange(int start, int end) {}

    public static StageRange getStageStepRange(int stage) {
        Integer end = LAST_STEP_INDEX_BY_STAGE.get(stage);
        if (end == null || end < 0) return null;
        int prevEnd = stage > 1 ? LAST_STEP_INDEX_BY_STAGE.getOrDefault(stage - 1, -1) : -1;
        return new StageRange(prevEnd + 1, end);
    }

    public static Transition buildCompletionTransition(AppRow current) {
        int currentStep = current.currentStep();
        int currentStage = current.currentStage();
        int nextStepIndex = currentStep + 1;
        boolean stageBoundary = LAST_STEP_INDEX_BY_STAGE.getOrDefault(currentStage, -1) == currentStep;
        boolean isLastStepGlobal = nextStepIndex >= TOTAL_STEPS;

        String nextStatus = current.status();
        String nextSubstatus = current.workflowSubstatus();
        int nextStage = currentStage;

        if (isLastStepGlobal) {
            nextStatus = "COMPLETED";
            nextSubstatus = "DONE";
        } else if (stageBoundary) {
            nextStatus = "IN_PROGRESS";
            nextSubstatus = "REVIEW";
            nextStage = currentStage + 1;
        } else if (nextStepIndex == INTEGRATION_START_IDX) {
            nextStatus = "IN_PROGRESS";
            nextSubstatus = "INTEGRATION";
        } else {
            nextStatus = "IN_PROGRESS";
            if (!"INTEGRATION".equals(nextSubstatus)) nextSubstatus = "DRAFT";
        }
        return new Transition(nextStepIndex, nextStatus, nextSubstatus, nextStage);
    }

    public static Transition buildRollbackTransition(AppRow current) {
        int prevIndex = Math.max(0, current.currentStep() - 1);
        String nextSubstatus = current.workflowSubstatus();
        if ("REVIEW".equals(nextSubstatus) || "DONE".equals(nextSubstatus)) nextSubstatus = "DRAFT";
        return new Transition(prevIndex, "IN_PROGRESS", nextSubstatus, current.currentStage());
    }

    public static Transition buildReviewTransition(AppRow current, boolean approve) {
        if (approve) {
            String sub = current.currentStep() == INTEGRATION_START_IDX ? "INTEGRATION" : "DRAFT";
            return new Transition(current.currentStep(), "IN_PROGRESS", sub, current.currentStage());
        }
        int nextStage = Math.max(1, current.currentStage() - 1);
        int nextStep = LAST_STEP_INDEX_BY_STAGE.getOrDefault(nextStage, 0);
        return new Transition(nextStep, "IN_PROGRESS", "REVISION", nextStage);
    }
}
